openedx/core/djangoapps/course_roles/rules.py

```diff
@@ -0,0 +1,65 @@
+"""
+django-rules and Bridgekeeper rules for course_roles related features
+"""
+from bridgekeeper.rules import Rule
+
+from opaque_keys.edx.keys import CourseKey
+from xblock.core import XBlock
+
+from openedx.core.djangoapps.course_roles.api import get_all_user_permissions_for_a_course
+from openedx.core.djangoapps.course_roles.data import CourseRolesPermission
+from openedx.core.djangoapps.course_roles.toggles import use_permission_checks
+from openedx.core.djangoapps.content.course_overviews.models import CourseOverview
+from xmodule.course_block import CourseBlock
+from openedx.core.djangoapps.django_comment_common.models import Role
+from xmodule.error_block import ErrorBlock
+
+
+class HasPermissionRule(Rule):  # lint-amnesty, pylint: disable=abstract-method, missing-class-docstring
+    """
+    Rule to check if a user has a permission for a course,
+    including, if applicable, organization-wide permissions
+    and instance-wide permissions.
+    """
+    def __init__(self, permission: CourseRolesPermission):
+        self.permission = permission
+
+    def check(self, user, instance=None):
+        if not use_permission_checks():
+            return False
+        if not user.is_authenticated:
+            return False
+        if isinstance(instance, CourseKey):
+            course_key = instance
+        elif isinstance(instance, (CourseBlock, CourseOverview)):
+            course_key = instance.id
+        elif isinstance(instance, (ErrorBlock, XBlock)):
+            course_key = instance.scope_ids.usage_id.course_key
+        else:
+            course_key = CourseKey.from_string(str(instance))
+
+        return self.permission in get_all_user_permissions_for_a_course(user, course_key)
+
+
+class HasForumsRolesRule(Rule):  # lint-amnesty, pylint: disable=abstract-method, missing-class-docstring
+    """
+    Rule to check if a user has a forum role for a course.
+    """
+    def __init__(self, *roles):
+        self.role = roles
+
+    def check(self, user, instance=None):
```

Comment:
What happens in this method if instance is None?

Reply:
Considering that this rule will be executed to validate the permissions with the method user.has_perm(permission, course), I will modify this rule so that if a course is not passed when calling .has_perm, the permission will be validated for all courses, which would be the behavior expected by django.